rust/core-lib/src/plugins/rpc.rs

```diff
@@ -219,11 +244,42 @@ impl PluginBufferInfo {
 
 impl PluginUpdate {
     pub fn new<D>(view_id: ViewIdentifier, rev: u64, delta: D, new_len: usize,
-                  edit_type: String, author: String) -> Self
+                  new_line_count: usize, edit_type: String, author: String) -> Self
         where D: Into<Option<RopeDelta>>
     {
         let delta = delta.into();
-        PluginUpdate { view_id, delta, new_len, rev, edit_type, author }
+        PluginUpdate { view_id, delta, new_len, new_line_count, rev, edit_type, author }
+    }
+}
+
+// maybe this should be in xi_rope? has a strong resemblance to the various
+// concrete `Metric` types.
+impl TextUnit {
+    /// Converts an offset in some unit to a concrete byte offset. Returns
+    /// `None` if the input offset is out of bounds in its unit space.
+    pub fn resolve_offset<T: Borrow<Rope>>(&self, text: T, offset: usize) -> Option<usize> {
+        let text = text.borrow();
+        match *self {
+            TextUnit::Utf8 => {
+                if offset >= text.len() {
+                    None
+                } else {
+                    if text.is_codepoint_boundary(offset) {
+                        offset.into()
+                    } else {
+                        text.prev_codepoint_offset(offset).into()
+                    }
+                }
+            }
+            TextUnit::Line => {
+                let num_lines = text.measure::<LinesMetric>() + 1;
```

Comment:
I would either name this "max_line_number", or move the `+ 1` to the next line, probably preferring the latter.

Reply:
Again my understanding is that to actually get the number of lines in the rope I need the + 1? Seems like I'm probably confused.